ujit_iface.c

```diff
@@ -727,6 +773,10 @@ rb_ujit_init(struct rb_ujit_options *options)
     cUjitDisasmInsn = rb_struct_define_under(cUjitDisasm, "Insn", "address", "mnemonic", "op_str", NULL);
 #endif
 
+    if (RUBY_DEBUG && rb_ujit_opts.gen_stats) {
```

Comment:
This seems a bit hacky? Maybe we could make it so `runtime_counters_to_hash` returns Qnil when `gen_stats` is false? What do you think?

Reply:
The main thing this tries to do is to not register `at_exit` call when we are not generating stats. Maybe I can call `at_exit` from C instead of defining a constant here and checking it in Ruby.

Reply:
I think what I would do is always define the at_exit method, and when `runtime_counters_to_hash` returns `nil`, return early from the at_exit method. Could also rename `runtime_counters_to_hash` to `get_stat_counters`.

Reply:
I don't like the idea of having an at_exit callback in release build.

Reply:
Do you dislike the idea because of a specific policy of the Ruby core team, or because you're worried about the small impact it could have on performance of spinning up Ruby processes, or something else?

IMO having an at_exit that returns early seems nice and simple. It's also possible we might want to run other code in there later. I feel like optionally defining the at_exit or using a constant or another file makes the code more convoluted and harder to follow.

Reply:
I think over any practical reasons I just like the cleanliness of not adding the callback at all. Not a very good reason probably 😅 . Moving the printout code to a separate file has the benefit of making it easier to remove them when we do want to upstream. I will put a comment in there instead for that purpose.

Reply:
Ok cool. I don't think I'd even necessarily remove it when upstreaming unless they specifically ask us to. It's a valuable tool after all for anyone who maintains, debugs or continues to develop uJIT.

Reply:
It looks like always having the at_exit is causing MJIT tests to fail:
```
2021-03-02T20:45:34.8612084Z  24) Failure:
2021-03-02T20:45:34.8612674Z TestJIT#test_compile_insn_invokeblock_leave [/home/runner/work/ruby/ruby/src/test/ruby/test_jit.rb:433]:
2021-03-02T20:45:34.8613400Z Expected 2 times of JIT success, but succeeded 4 times.
2021-03-02T20:45:34.8613769Z 
2021-03-02T20:45:34.8614051Z script:
2021-03-02T20:45:34.8614317Z """
2021-03-02T20:45:34.8614500Z 
2021-03-02T20:45:34.8614897Z def foo
2021-03-02T20:45:34.8615189Z   yield
2021-03-02T20:45:34.8615459Z end
2021-03-02T20:45:34.8615743Z print foo { 2 }
2021-03-02T20:45:34.8615968Z 
2021-03-02T20:45:34.8616225Z """
2021-03-02T20:45:34.8616394Z 
2021-03-02T20:45:34.8616577Z 
2021-03-02T20:45:34.8616839Z stderr:
2021-03-02T20:45:34.8617116Z """
2021-03-02T20:45:34.8617664Z JIT success (2345.3ms): foo@-e:2 -> /tmp/_ruby_mjit_p31553u0.c
2021-03-02T20:45:34.8618370Z JIT success (38.4ms): block in <main>@-e:5 -> /tmp/_ruby_mjit_p31553u1.c
2021-03-02T20:45:34.8619290Z JIT success (219.5ms): block in <module:UJIT>@<internal:ujit>:46 -> /tmp/_ruby_mjit_p31553u2.c
2021-03-02T20:45:34.8620158Z JIT success (80.1ms): runtime_stats@<internal:ujit>:41 -> /tmp/_ruby_mjit_p31553u3.c
2021-03-02T20:45:34.8620760Z Successful MJIT finish
2021-03-02T20:45:34.8621025Z 
```

Reply:
😢 Ok, we can go with your original solution with the constant then.

Reply:
Or call Ruby from C, as you prefer.